base/utf.jl

```diff
@@ -0,0 +1,880 @@
+    len = length(str)
+    buf = Vector{Char}(len+1)
+    out = 0
+    @inbounds for ch in str ; buf[out += 1] = ch ; end
+    buf[out += 1] = Char(0) # NULL termination
+    UTF32String(buf)
+end
+
+#=
+@doc """
+@brief      Converts a UTF-8 encoded string to UTF-16 encoding
+
+@param[in]  src::Vector{UInt8}
+
+@return     ::UTF16String
+@throws     ArgumentError
+""" ->
+=#
+function convert(::Type{UTF16String}, src::UTF8String)
+    str = src.data
+    # handle zero length string quickly
+    if sizeof(str) == 0 ; buf = Vector{UInt16}(1) ; buf[1] = 0 ; return UTF16String(buf) ; end
+    # Check that is correct UTF-8 encoding and get number of words needed
+    cnts::UTF_String_Counts = check_string_utf8(str)
+    len = cnts.cntT + cnts.cnt4
+    buf = Vector{UInt16}(len+1)
+    buf[len+1] = 0
+    # Optimize case where no characters > 0x7f
+    cnts.flags == 0 && return UTF16String(copy!(buf, str))
+    out::UInt = 0
+    pos::UInt = 0
+    @inbounds while out < len
+        ch::UInt32 = str[pos += 1]
+        # Handle ASCII characters
+        if ch <= 0x7f
+            buf[out += 1] = ch
+        # Handle range 0x80-0x7ff
+        elseif ch < 0xe0
+            buf[out += 1] = @get_utf8_2!(str, pos, ch)
+        # Handle range 0x800-0xffff
+        elseif ch < 0xf0
+            buf[out += 1] = @get_utf8_3!(str, pos, ch)
+        # Handle range 0x10000-0x10ffff
+        else
+            ch = @get_utf8_4!(str, pos, ch)
+            @output_utf16_surr!(buf, out, ch)
+        end
+    end
+    UTF16String(buf)
+end
+
+#=
+@doc """
+@brief      Reencodes a UTF-16 or UTF-32 encoded string using UTF-8 encoding
+
+@param[in]  str::Union(Vector{UInt16}, Vector{UInt32})
+
+@return     ::UTF8String
+@throws     ArgumentError
+""" ->
+=#
+function convert(::Type{UTF8String}, str::Vector{UInt16})
+    len = sizeof(str)
+    # handle zero length string quickly
+    len == 0 && return UTF8String("")
+    # get number of bytes to allocate
+    cnts::UTF_String_Counts = check_string_utf16(str, len>>>1)
+    cnts.flags == 0 && return UTF8String(copy!(Vector{UInt8}(cnts.cntT), 1, str, 1, cnts.cntT))
+    return encode_to_8(UInt16, str, cnts.cntT + cnts.cnt2 + cnts.cnt3*2 + cnts.cnt4*3)
+end
+
+function convert(::Type{UTF8String}, src::UTF16String)
+    str = src.data
+    len = sizeof(str) >>> 1
+    # handle zero length string quickly
+    len <= 1 && return UTF8String("")
+    # get number of bytes to allocate
+    cnts::UTF_String_Counts = check_string_utf16(str, len-1)
+    cnts.flags == 0 && return UTF8String(copy!(Vector{UInt8}(cnts.cntT), 1, str, 1, cnts.cntT))
+    return encode_to_8(UInt16, str, cnts.cntT + cnts.cnt2 + cnts.cnt3*2 + cnts.cnt4*3)
+end
+
+function convert(::Type{UTF8String}, str::Vector{UInt32})
+    len = sizeof(str)
+    # handle zero length string quickly
+    len == 0 && return UTF8String("")
+    # get number of bytes to allocate
+    cnts::UTF_String_Counts = check_string_utf32(str, len>>>2)
```

Comment:
Just unfamiliarity with Julia, I'll change those.  Thanks very much for the thorough review!